community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java

```diff
@@ -312,7 +316,7 @@ public void markForTermination()
             if ( stillSameTransaction && canBeTerminated() )
             {
                 failure = true;
-                terminated = true;
+                terminationReason = reason;
```

Comment:
There is another problem where a tx that is being `markForTermination` while being `close` and the termination is not scheduled until the tx has returned to the pool and then initialised again, effectively terminating a completely different tx.

Here is a branch with test that reproduce this problem in a single thread by keeping the reference to the old `TopLevelTransaction` which holds the pooled `KernelTransactionImplementation`.
https://github.com/burqen/neo4j/tree/2.3-tx-terminate-after-reuse

I suggest grabbing the lock earlier in `markForTermination` and otherwise leave this problem to be adressed separate.